File: androidApp/src/main/kotlin/com/virginactive/android/ui/detail/ClassDetailScreen.kt
package com.virginactive.android.ui.detail

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.AlertDialog
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Scaffold
import androidx.compose.material3.SnackbarHost
import androidx.compose.material3.SnackbarHostState
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.TopAppBar
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.virginactive.android.theme.CardCornerRadius
import com.virginactive.android.theme.OnWarningContainer
import com.virginactive.android.theme.Spacing
import com.virginactive.android.theme.WarningBorder
import com.virginactive.android.theme.WarningContainer
import com.virginactive.android.ui.common.ErrorState
import com.virginactive.android.ui.common.LoadingState
import com.virginactive.shared.domain.booking.ClassInstance
import com.virginactive.shared.domain.booking.ClassStatus
import com.virginactive.shared.domain.booking.UserBookingStatus
import com.virginactive.shared.domain.home.DateTimeDisplay
import org.koin.androidx.compose.koinViewModel

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun ClassDetailScreen(
    classId: String,
    onBack: () -> Unit,
    modifier: Modifier = Modifier,
    snackbarHostState: SnackbarHostState = remember { SnackbarHostState() },
    vm: ClassDetailViewModel = koinViewModel(),
) {
    LaunchedEffect(classId) { vm.load(classId) }

    val state = vm.state

    LaunchedEffect(state.actionError) {
        if (state.actionError == null) return@LaunchedEffect
        snackbarHostState.showSnackbar("We couldn't complete that. Tap to try again.")
        vm.actionErrorShown()
    }

    Scaffold(
        modifier = modifier,
        snackbarHost = { SnackbarHost(snackbarHostState) },
        topBar = {
            TopAppBar(
                title = { Text("Class Details") },
                navigationIcon = {
                    TextButton(onClick = onBack) { Text("Back") }
                },
            )
        },
    ) { innerPadding ->
        when {
            state.isLoading -> LoadingState(modifier = Modifier.padding(innerPadding))

            state.error != null -> ErrorState(
                onRetry = { vm.load(classId) },
                modifier = Modifier.padding(innerPadding),
            )

            state.classInstance != null -> ClassDetailBody(
                instance = state.classInstance,
                confirmation = state.confirmation,
                actionInFlight = state.actionInFlight,
                withinForfeitWindow = vm.isWithinForfeitWindow(),
                onBook = vm::book,
                onCancel = vm::cancel,
                snackbarHostState = snackbarHostState,
                modifier = Modifier.padding(innerPadding),
            )
        }
    }
}

@Composable
private fun ClassDetailBody(
    instance: ClassInstance,
    confirmation: Confirmation?,
    actionInFlight: Boolean,
    withinForfeitWindow: Boolean,
    onBook: () -> Unit,
    onCancel: () -> Unit,
    snackbarHostState: SnackbarHostState,
    modifier: Modifier = Modifier,
) {
    var showReminderSheet by remember { mutableStateOf(false) }
    var showCancelDialog by remember { mutableStateOf(false) }
    var reminderMessage by remember { mutableStateOf<String?>(null) }
    var reminderAction by remember { mutableStateOf<(() -> Unit)?>(null) }

    LaunchedEffect(reminderMessage) {
        val message = reminderMessage ?: return@LaunchedEffect
        val result = snackbarHostState.showSnackbar(message, actionLabel = "Open settings")
        if (result == androidx.compose.material3.SnackbarResult.ActionPerformed) reminderAction?.invoke()
        reminderMessage = null
        reminderAction = null
    }

    val isBooked = confirmation?.userBookingStatus == UserBookingStatus.BOOKED ||
        confirmation?.userBookingStatus == UserBookingStatus.WAITLISTED

    Column(
        modifier = modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(horizontal = Spacing.md, vertical = Spacing.md),
    ) {
        TypeChip(type = instance.type.name)
        Spacer(Modifier.height(Spacing.sm))
        Text(
            text = instance.title,
            style = MaterialTheme.typography.headlineSmall,
            color = MaterialTheme.colorScheme.onSurface,
        )
        Text(
            text = "with ${instance.trainer}",
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            modifier = Modifier.padding(top = Spacing.xs),
        )

        Spacer(Modifier.height(Spacing.md))
        InfoCard(instance = instance)

        if (confirmation != null) {
            Spacer(Modifier.height(Spacing.md))
            ConfirmationContent(confirmation = confirmation)
        }

        if (withinForfeitWindow && isBooked) {
            Spacer(Modifier.height(Spacing.md))
            ForfeitStrip()
        }

        Spacer(Modifier.height(Spacing.lg))
        PrimaryAction(
            status = instance.status,
            actionInFlight = actionInFlight,
            onBook = onBook,
        )

        if (isBooked) {
            Spacer(Modifier.height(Spacing.sm))
            Button(
                onClick = { showReminderSheet = true },
                modifier = Modifier.fillMaxWidth(),
                colors = ButtonDefaults.buttonColors(
                    containerColor = MaterialTheme.colorScheme.surfaceVariant,
                    contentColor = MaterialTheme.colorScheme.onSurface,
                ),
            ) {
                Text("Set Reminder")
            }

            Spacer(Modifier.height(Spacing.sm))
            TextButton(
                onClick = { showCancelDialog = true },
                modifier = Modifier.fillMaxWidth(),
            ) {
                Text("Cancel Booking", color = MaterialTheme.colorScheme.error)
            }
        }
    }

    if (showReminderSheet) {
        ReminderBottomSheet(
            classInstance = instance,
            onDismiss = { showReminderSheet = false },
            onSnackbar = { message, _, onAction ->
                reminderMessage = message
                reminderAction = onAction
            },
        )
    }

    if (showCancelDialog) {
        CancelDialog(
            withinForfeitWindow = withinForfeitWindow,
            onConfirm = {
                showCancelDialog = false
                onCancel()
            },
            onDismiss = { showCancelDialog = false },
        )
    }
}

@Composable
private fun TypeChip(type: String) {
    Text(
        text = type,
        style = MaterialTheme.typography.labelLarge,
        color = MaterialTheme.colorScheme.onPrimary,
        modifier = Modifier
            .clip(RoundedCornerShape(percent = 50))
            .background(MaterialTheme.colorScheme.primary)
            .padding(horizontal = Spacing.sm, vertical = Spacing.xs),
    )
}

@Composable
private fun InfoCard(instance: ClassInstance) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(CardCornerRadius))
            .background(MaterialTheme.colorScheme.surfaceVariant)
            .padding(Spacing.md),
    ) {
        val offset = DateTimeDisplay.offsetLabel(instance.startsAt)
        InfoRow(label = "Date", value = DateTimeDisplay.dateLabel(instance.startsAt))
        InfoRow(
            label = "Time",
            value = buildString {
                append(DateTimeDisplay.timeLabel(instance.startsAt))
                append(" – ")
                append(DateTimeDisplay.timeLabel(instance.endsAt))
                if (offset.isNotEmpty()) append(" ($offset)")
            },
        )
        InfoRow(label = "Availability", value = "${instance.available} of ${instance.spots} spots")
    }
}

@Composable
private fun InfoRow(label: String, value: String) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .padding(vertical = Spacing.xs),
        horizontalArrangement = Arrangement.SpaceBetween,
    ) {
        Text(
            text = label,
            style = MaterialTheme.typography.bodyMedium,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
        )
        Text(
            text = value,
            style = MaterialTheme.typography.labelLarge,
            color = MaterialTheme.colorScheme.onSurface,
        )
    }
}

@Composable
private fun ForfeitStrip() {
    Box(
        modifier = Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(CardCornerRadius))
            .background(WarningContainer)
            .border(width = 2.dp, color = WarningBorder, shape = RoundedCornerShape(CardCornerRadius))
            .padding(Spacing.md),
    ) {
        Text(
            text = "Cancelling within 12 hours of the class incurs a forfeit.",
            style = MaterialTheme.typography.bodyMedium,
            color = OnWarningContainer,
        )
    }
}

@Composable
private fun PrimaryAction(
    status: ClassStatus,
    actionInFlight: Boolean,
    onBook: () -> Unit,
) {
    val label = if (status == ClassStatus.FULL) "Join Waitlist" else "Book Class"
    Button(
        onClick = onBook,
        enabled = !actionInFlight && status != ClassStatus.CANCELLED,
        modifier = Modifier.fillMaxWidth(),
    ) {
        if (actionInFlight) {
            CircularProgressIndicator(
                modifier = Modifier.size(20.dp),
                color = MaterialTheme.colorScheme.onPrimary,
                strokeWidth = 2.dp,
            )
        } else {
            Text(label)
        }
    }
}

@Composable
private fun CancelDialog(
    withinForfeitWindow: Boolean,
    onConfirm: () -> Unit,
    onDismiss: () -> Unit,
) {
    AlertDialog(
        onDismissRequest = onDismiss,
        title = { Text("Cancel this booking?") },
        text = {
            Text(
                if (withinForfeitWindow) {
                    "This class starts within 12 hours, so cancelling now incurs a forfeit."
                } else {
                    "You can rebook later if spots are available."
                },
            )
        },
        confirmButton = {
            TextButton(onClick = onConfirm) {
                Text("Cancel Booking", color = MaterialTheme.colorScheme.error)
            }
        },
        dismissButton = {
            TextButton(onClick = onDismiss) { Text("Keep Booking") }
        },
    )
}

File: iosApp/iosApp/iosApp/ClassDetail/ClassDetailScreen.swift
import SwiftUI
import Shared

struct ClassDetailScreen: View {
    let classId: String

    @StateObject private var vm: ClassDetailViewModel

    init(component: AppComponent, classId: String) {
        self.classId = classId
        _vm = StateObject(wrappedValue: ClassDetailViewModel(component: component))
    }

    var body: some View {
        content
            .navigationTitle("Class Details")
            .navigationBarTitleDisplayMode(.inline)
            .task { vm.load(classId: classId) }
            .onDisappear { vm.cancelTasks() }
    }

    @ViewBuilder
    private var content: some View {
        if vm.isLoading {
            LoadingView()
        } else if vm.error != nil {
            ErrorView(retry: { vm.load(classId: classId) })
        } else if let instance = vm.classInstance {
            ClassDetailBody(
                instance: instance,
                confirmation: vm.confirmation,
                actionInFlight: vm.actionInFlight,
                withinForfeitWindow: vm.isWithinForfeitWindow(),
                actionErrorMessage: vm.actionErrorMessage,
                onBook: vm.book,
                onCancel: vm.cancel,
                onActionErrorShown: vm.actionErrorShown
            )
        } else {
            EmptyStateView(heading: "Nothing to show yet")
        }
    }
}

private struct ClassDetailBody: View {
    let instance: ClassInstance
    let confirmation: Confirmation?
    let actionInFlight: Bool
    let withinForfeitWindow: Bool
    let actionErrorMessage: String?
    let onBook: () -> Void
    let onCancel: () -> Void
    let onActionErrorShown: () -> Void

    @State private var showReminderSheet = false
    @State private var showCancelAlert = false
    @State private var toastMessage: String?

    private var isBooked: Bool {
        confirmation?.userBookingStatus == .booked || confirmation?.userBookingStatus == .waitlisted
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                TypeChip(type: instance.type)

                Text(instance.title)
                    .font(.vaDisplay)
                    .foregroundStyle(VaColor.onSurface)
                    .padding(.top, VaSpacing.sm)

                Text("with \(instance.trainer)")
                    .font(.vaBody)
                    .foregroundStyle(VaColor.onSurfaceVariant)
                    .padding(.top, VaSpacing.xs)

                InfoCard(instance: instance)
                    .padding(.top, VaSpacing.md)

                if let confirmation {
                    ConfirmationView(confirmation: confirmation)
                        .padding(.top, VaSpacing.md)
                }

                if withinForfeitWindow && isBooked {
                    ForfeitStrip()
                        .padding(.top, VaSpacing.md)
                }

                PrimaryAction(status: instance.status, actionInFlight: actionInFlight, onBook: onBook)
                    .padding(.top, VaSpacing.lg)

                if isBooked {
                    Button(action: { showReminderSheet = true }) {
                        Text("Set Reminder")
                            .font(.vaLabel)
                            .foregroundStyle(VaColor.onSurface)
                            .frame(maxWidth: .infinity, minHeight: VaSpacing.minTouchTarget)
                            .background(VaColor.surfaceVariant)
                            .clipShape(RoundedRectangle(cornerRadius: VaSpacing.cardCornerRadius))
                    }
                    .padding(.top, VaSpacing.sm)

                    Button(role: .destructive, action: { showCancelAlert = true }) {
                        Text("Cancel Booking")
                            .font(.vaLabel)
                            .foregroundStyle(VaColor.errorRed)
                            .frame(maxWidth: .infinity, minHeight: VaSpacing.minTouchTarget)
                    }
                    .padding(.top, VaSpacing.sm)
                }
            }
            .padding(.horizontal, VaSpacing.md)
            .padding(.vertical, VaSpacing.md)
        }
        .toast($toastMessage)
        .onChange(of: actionErrorMessage) { _, newValue in
            guard let newValue else { return }
            toastMessage = newValue
            onActionErrorShown()
        }
        .alert("Cancel this booking?", isPresented: $showCancelAlert) {
            Button("Cancel Booking", role: .destructive, action: onCancel)
            Button("Keep Booking", role: .cancel) {}
        } message: {
            Text(
                withinForfeitWindow
                    ? "This class starts within 12 hours, so cancelling now incurs a forfeit."
                    : "You can rebook later if spots are available."
            )
        }
        .sheet(isPresented: $showReminderSheet) {
            ReminderSheet(classInstance: instance) { message, _ in
                toastMessage = message
                showReminderSheet = false
            } onScheduled: {
                showReminderSheet = false
            }
        }
    }
}

private struct TypeChip: View {
    let type: ClassType

    private var label: String {
        switch type {
        case .groupWorkout: return "GROUP WORKOUT"
        case .yoga: return "YOGA"
        case .spin: return "SPIN"
        case .pilates: return "PILATES"
        case .hiit: return "HIIT"
        case .swimming: return "SWIMMING"
        case .unknown: return "CLASS"
        @unknown default: return "CLASS"
        }
    }

    var body: some View {
        Text(label)
            .font(.vaLabel)
            .foregroundStyle(VaColor.white)
            .padding(.horizontal, VaSpacing.sm)
            .padding(.vertical, VaSpacing.xs)
            .background(VaColor.vaRed)
            .clipShape(Capsule())
    }
}

private struct InfoCard: View {
    let instance: ClassInstance

    private var timeValue: String {
        let display = DateTimeDisplay.shared
        let offset = display.offsetLabel(isoWithOffset: instance.startsAt)
        var value = "\(display.timeLabel(isoWithOffset: instance.startsAt)) – \(display.timeLabel(isoWithOffset: instance.endsAt))"
        if !offset.isEmpty { value += " (\(offset))" }
        return value
    }

    var body: some View {
        VStack(spacing: 0) {
            InfoRow(label: "Date", value: DateTimeDisplay.shared.dateLabel(isoWithOffset: instance.startsAt))
            InfoRow(label: "Time", value: timeValue)
            InfoRow(label: "Availability", value: "\(instance.available) of \(instance.spots) spots")
        }
        .frame(maxWidth: .infinity)
        .padding(VaSpacing.md)
        .background(VaColor.surfaceVariant)
        .clipShape(RoundedRectangle(cornerRadius: VaSpacing.cardCornerRadius))
    }
}

private struct InfoRow: View {
    let label: String
    let value: String

    var body: some View {
        HStack {
            Text(label)
                .font(.vaBody)
                .foregroundStyle(VaColor.onSurfaceVariant)
            Spacer()
            Text(value)
                .font(.vaLabel)
                .foregroundStyle(VaColor.onSurface)
                .multilineTextAlignment(.trailing)
        }
        .padding(.vertical, VaSpacing.xs)
    }
}

private struct ForfeitStrip: View {
    var body: some View {
        Text("Cancelling within 12 hours of the class incurs a forfeit.")
            .font(.vaBody)
            .foregroundStyle(VaColor.onWarningContainer)
            .frame(maxWidth: .infinity, alignment: .leading)
            .padding(VaSpacing.md)
            .background(VaColor.warningContainer)
            .clipShape(RoundedRectangle(cornerRadius: VaSpacing.cardCornerRadius))
            .overlay(
                RoundedRectangle(cornerRadius: VaSpacing.cardCornerRadius)
                    .stroke(VaColor.warningBorder, lineWidth: 2)
            )
    }
}

private struct PrimaryAction: View {
    let status: ClassStatus
    let actionInFlight: Bool
    let onBook: () -> Void

    private var label: String { status == .full ? "Join Waitlist" : "Book Class" }

    var body: some View {
        Button(action: onBook) {
            ZStack {
                if actionInFlight {
                    ProgressView()
                        .progressViewStyle(.circular)
                        .tint(VaColor.white)
                } else {
                    Text(label).font(.vaLabel)
                }
            }
            .foregroundStyle(VaColor.white)
            .frame(maxWidth: .infinity, minHeight: VaSpacing.minTouchTarget)
            .background(status == .cancelled ? VaColor.vaRed.opacity(0.4) : VaColor.vaRed)
            .clipShape(RoundedRectangle(cornerRadius: VaSpacing.cardCornerRadius))
        }
        .disabled(actionInFlight || status == .cancelled)
    }
}
